File: android/app/src/main/java/com/phantomcatworks/okusuritechou/data/db/entity/NotificationSetting.kt
package com.phantomcatworks.okusuritechou.data.db.entity

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.phantomcatworks.okusuritechou.data.db.AppHolder
import kotlinx.coroutines.flow.Flow

@Entity(tableName = "notification_settings")
data class NotificationSetting(
    @PrimaryKey val slot: TimingSlot,
    val hour: Int,
    val minute: Int,
    val message: String,
    val enabled: Boolean = true
) {
    suspend fun hasNoApplicableMedicine(): Boolean {
        val actives = Medicine.active()
        return actives.none {
            when (slot) {
                TimingSlot.morning -> it.timingMorning
                TimingSlot.noon -> it.timingNoon
                TimingSlot.night -> it.timingNight
            }
        }
    }

    suspend fun save(): NotificationSetting {
        AppHolder.db.notificationSettingDao().upsert(this)
        return this
    }

    companion object {
        private val dao get() = AppHolder.db.notificationSettingDao()

        val DEFAULTS = listOf(
            NotificationSetting(TimingSlot.morning, 8, 0, "朝のお薬の時間です。お忘れなく服用しましょう。"),
            NotificationSetting(TimingSlot.noon, 12, 0, "昼のお薬の時間です。お忘れなく服用しましょう。"),
            NotificationSetting(TimingSlot.night, 18, 0, "夜のお薬の時間です。お忘れなく服用しましょう。")
        )

        fun observeAll(): Flow<List<NotificationSetting>> = dao.observeAll()
        suspend fun all(): List<NotificationSetting> = dao.findAll()
        suspend fun find(slot: TimingSlot): NotificationSetting? = dao.findBySlot(slot)

        suspend fun ensureDefaults() {
            DEFAULTS.forEach { default ->
                if (dao.findBySlot(default.slot) == null) {
                    dao.upsert(default)
                }
            }
        }
    }
}

File: ios/OkusuriTechou/Data/Models/NotificationSetting.swift
import Foundation
import GRDB

struct NotificationSetting: Codable, FetchableRecord, PersistableRecord {
    static let databaseTableName = "notification_settings"

    var slot: TimingSlot
    var hour: Int
    var minute: Int
    var message: String
    var enabled: Bool

    static let defaults: [NotificationSetting] = [
        NotificationSetting(slot: .morning, hour: 8, minute: 0, message: "朝のお薬の時間です。お忘れなく服用しましょう。", enabled: true),
        NotificationSetting(slot: .noon, hour: 12, minute: 0, message: "昼のお薬の時間です。お忘れなく服用しましょう。", enabled: true),
        NotificationSetting(slot: .night, hour: 18, minute: 0, message: "夜のお薬の時間です。お忘れなく服用しましょう。", enabled: true)
    ]

    func hasNoApplicableMedicine() throws -> Bool {
        let actives = try Medicine.active()
        return !actives.contains { medicine in
            switch slot {
            case .morning: return medicine.timingMorning
            case .noon: return medicine.timingNoon
            case .night: return medicine.timingNight
            }
        }
    }

    @discardableResult
    func save() throws -> NotificationSetting {
        try AppDatabaseManager.shared.write { db in try self.save(db) }
        return self
    }

    static func all() throws -> [NotificationSetting] {
        try AppDatabaseManager.shared.read { db in try NotificationSetting.fetchAll(db) }
    }

    static func find(_ slot: TimingSlot) throws -> NotificationSetting? {
        try AppDatabaseManager.shared.read { db in try NotificationSetting.fetchOne(db, key: slot.rawValue) }
    }

    static func ensureDefaults() throws {
        try AppDatabaseManager.shared.write { db in
            for setting in defaults {
                if try NotificationSetting.fetchOne(db, key: setting.slot.rawValue) == nil {
                    try setting.insert(db)
                }
            }
        }
    }
}
